pkg/container/queue/chunkqueue.go

```diff
@@ -0,0 +1,389 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package queue
+
+import (
+	"sync"
+	"unsafe"
+
+	"github.com/pingcap/log"
+)
+
+const (
+	// the size of each chunk is 1024 bytes (1kB) by default
+	defaultSizePerChunk = 1024
+	// the minimum length of each chunk is 16
+	minimumChunkLen      = 16
+	defaultPitchArrayLen = 16
+)
+
+// ChunkQueue is a generic, efficient, iterable and GC-friendly queue.
+// Attention, it's not thread-safe.
+type ChunkQueue[T any] struct {
+	// [head, tail) is the section of chunks in use
+	head int
+	tail int
+
+	// size is number of elements in queue
+	size int
+
+	// chunks is an array storing ptr
+	chunks []*chunk[T]
+	// chunkLength is the max number of elements stored in every chunk
+	chunkLength  int
+	chunkPool    sync.Pool
+	defaultValue T
+}
+
+func (q *ChunkQueue[T]) firstChunk() *chunk[T] {
+	return q.chunks[q.head]
+}
+
+func (q *ChunkQueue[T]) lastChunk() *chunk[T] {
+	return q.chunks[q.tail-1]
+}
+
+// NewChunkQueue creates a new ChunkQueue
+func NewChunkQueue[T any]() *ChunkQueue[T] {
+	return NewChunkQueueLeastCapacity[T](1)
+}
+
+// NewChunkQueueLeastCapacity creates a ChunkQueue with an argument minCapacity.
+// It requests that the queue capacity be at least minCapacity. And it's similar
+// to the cap argument when making a slice using make([]T, len, cap)
+func NewChunkQueueLeastCapacity[T any](minCapacity int) *ChunkQueue[T] {
+	elementSize := unsafe.Sizeof(*new(T))
```

Comment:
If we define `type E struct{}`, then the size of a E variable is zero. 
In fact, it can also be supported. The reasons it's not supported are 1) to avoid zero division in calculating `chunkLength`, and 2) scenarios of using it can be handled well by channels